Fix pipewire buffer processing

mbernat opened the pull request:
## Description

This is a more intense PR that attempts to fix several issues with pipewire buffer processing. It does multiple things to solve multiple issues and it might be controversial, so please consider it just a conversation starting point, I'd be happy to cut it up into multiple PRs and/or use alternative approaches, whatever is deemed best.

- Chunk sizes were not respected and the whole scratch buffer was used instead.
  - On the input side pipewire tells us how much data was captured and we should respect that. I _think_ we should also not write into the chunk data here (that's reserved for telling pipewire how much data _we_ are providing on the output side below).
  - On the output side there is `Buffer::requested()` API to tell us how much data is expected in this period but it's only available in 0.3.49. This is higher than the Ubuntu LTS version (0.3.48), so we should find a fall back when using an older version. There are many other things we could do here: use the sizes that the user provided in the stream config or query the device properties for some defaults.
- Putting multiple chunks to and from the scratch buffer did not work as expected.
  - I believe the old frame processor wrongly reordered data in case of multiple channels and multiple chunks. I solved it by switching to interleaved sampling (`F32LE`), which trivializes the problem: in this case the chunks can simply be joined. But it would also be possible to keep the non-interleaved sampling (`F32P`) and reorder the channel data correctly.
- Removed from this PR but closely related to the above: we would like to use fixed buffer sizes for arcane AEC webrtc processor reasons. I found that it is possible to use [node.force-quantum](https://docs.pipewire.org/group__pw__keys.html#gaa50ce0bca22904a417a4e9d05f8face6) to achieve this but I do not know if this is the correct way to approach things. I suppose there might be softer ways to tell pipewire that you'd like to use buffer sizes in a given range.

## Type of Change

Using the `Buffer::requested()` API is a breaking change but maybe it's not strictly necessary. The rest of the PR should be non-breaking.

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How Has This Been Tested?

- [ ] The `sine_wave` example did not work for me as expected, it was choppy and the callback was only called few times a second (presumably because the 8192 scratch buffer was provided every time, so pipewire decided 6 calls are enough to get the required 48k ~ 6 * 4192 samples). With this PR it's using buffer sizes proportional to the default device quantization (1024 in my case), gets called roughly every 20ms, and the expected sine wave sound is produced.
- [ ] Adding callback timestamp info to the `input` example reveals that it suffered from similar problems, it was only called few times a second. I haven't recorded audio samples but I suspect they would have been garbled. After this PR audio capture works fine on a variety of devices (mics and capture card).
- [ ] The processing of multiple chunks is harder to test, pipewire typically uses just 1 chunk, but I did run into it on some devices and before this PR I got first crashes and then garbled input/output, with the PR in its current form it seems to work fine.

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] Wherever possible, I have added tests that prove my fix is effective or that my feature works. For changes that
      need to be validated manually (i.e. a new audio driver), use examples that can be run to easily validate them.
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

## Screenshots (if appropriate):

## Additional Notes:

Add any additional notes about the PR here.

mbernat committed:
Fix pipewire backend

- Provide requested number of frames on the output
- Use chunk size to determine the number of frames on the input

mbernat committed:
Fix chunk processing by interleaving

SolarLiner reviewed (approved):
LGTM! Not only does it fix bugs, it also nicely cleans the code a little. Thanks!

mbernat commented on src/backends/pipewire/stream.rs:
```diff
@@ -190,26 +189,26 @@ impl<Callback: 'static + Send> StreamHandle<Callback> {
                         return;
                     }
                     if let Some(mut buffer) = stream.dequeue_buffer() {
+                        let requested_frames = if direction == Direction::Output {
+                            let requested_frames = buffer.requested() as usize;
+                            if requested_frames == 0 {
+                                log::warn!("0 frames were requested");
+                                return;
+                            }
+                            requested_frames
+                        } else {
+                            0
+                        };
```
What do you mean by the "given" sizes?
- We could use the buffer size range from the stream config but that could be unset and it's also unclear whether we'd use the lower or the upper limit or some combination of the two, if both are set.
- Or we could query the device for the default size information.
- But at the moment I don't think there is such a thing as a "given" size, which is why I reached for the `requested` API. That being said, there might very well be other options.

As for your other questions I wasn't quite sure, so I decided to retest the behavior.
- Pipewire expects to process a certain number of frames in a given period so that it can hit the target sample rate.
- If you provide the requested number of frames then everything behaves extremely smoothly, for example with 48k sample rate and 1024 frames you get called roughly every 21ms.
- If you provide a multiple of the requested number you just get called less often but things are still smooth, for example every 43ms providing 2048 frames.
- If you provide too many samples (above 4096 in my case) the audio starts getting choppy.
- If you provide a larger non-multiple then you get called less consistently, for example it's sometimes 21ms and sometimes 43ms when providing 1300 frames.
- If you provide fewer frames then it's even less consistent: for example, one 18ms call and 3 1ms calls when providing 256 frames.
- There seems to be a limit on how many times pipewire calls you, so if you provide too few frames (16 or fewer in my case) then pipewire cannot hit the target sample rate anymore and the audio breaks down.
- Pipewire gradually adjusts the number of requested frames if you don't provide the requested number.
  ```
  Requested 1024
  Elapsed: 17.815
  Requested 768
  Elapsed: 1.326
  Requested 512
  Elapsed: 1.183
  Requested 256
  Elapsed: 1.191
  ```

- The upshot of all of the above is that pipewire actually works sounds great in a large range of buffer sizes, so we could probably get away with hard coding a size like 512 or 1024 as a reasonable default.
- It would be good to use the native device size (like 1024 above) to get the optimal latency and consistency but I'm not sure how to obtain this number. `pw-top` shows it after I create a stream but I don't see it in `wpctl` properties.
- Alternatively, we can configure the stream to use a specific size with `node.force-quantum`. Pipewire then requests this number of frames and runs optimally if we provide that number of frames.

I'm leaning towards the last approach by using the buffer size range from the stream config and defaulting to 512 otherwise (it's a nice number and CoreAudio also happens to use it as a default, so we'd be in a good company).

SolarLiner replied:
I have put 512 as a default in CoreAudio, more because I was waiting for a better solution down the line, probably also looking at how to query for a default buffer size from the platform (if that even is consistently possible). The buffer size range being possibly completely open (meaning that the user does not have any preference in latency or processing size) means that we can technically do what we want in that case, and I'm not too against having platform-specific behavior in choosing the actual buffer size if it means better performance in one way or another.

Because there is a single chosen buffer size that is "resolved" when starting the stream, once that number is chosen, the user callbacks should always be called with the same buffer size as well, so this limits what we can do when a stream is already running. We can add a way to notify the buffer size has changed, but I think a common expectation with audio callbacks is that they don't actually change size (that trips a number of people with cpal which IIRC forwards however many frames the platform has requested/provided). We can also eventually make this behavior configurable, where the audio callback provides a callback configuration of its own (and that's where we could let the user say that they do support variable-size buffers). For now, for simplicity's sake I think it's better to keep one size and stick to it.

I think from what I understand of your message, that we should therefore make use of `node.force-quantum` all of the time, then, and go through similar heuristics as other platforms to choose that number before the stream starts (i.e., in CoreAudio it's min size, or else max size, or else 512 as default)

mbernat committed:
Stop using the `Buffer::requested()` API

- Use a fixed buffer size using the `node.force-quantum` stream property instead
- The size is computed from the stream config buffer size range and defaulting to 512
- With this the pipewire version can be reduced to 0.3.45

mbernat commented:
I have implemented the tl;dr from the comment, let me know what you think!

mbernat commented:
I'm not sure what's wrong with the CI. Maybe temporary network issues?

```
E: Failed to fetch https://security.ubuntu.com/ubuntu/pool/main/p/pipewire/libspa-0.2-modules_1.0.5-1ubuntu3_amd64.deb  404  Not Found [IP: 52.252.75.106 80]
E: Failed to fetch https://security.ubuntu.com/ubuntu/pool/main/p/pipewire/libpipewire-0.3-0t64_1.0.5-1ubuntu3_amd64.deb  404  Not Found [IP: 52.252.75.106 80]
E: Failed to fetch https://security.ubuntu.com/ubuntu/pool/main/p/pipewire/libpipewire-0.3-common_1.0.5-1ubuntu3_all.deb  404  Not Found [IP: 52.252.75.106 80]
E: Failed to fetch https://security.ubuntu.com/ubuntu/pool/main/p/pipewire/libspa-0.2-dev_1.0.5-1ubuntu3_amd64.deb  404  Not Found [IP: 52.252.75.106 80]
E: Failed to fetch https://security.ubuntu.com/ubuntu/pool/main/p/pipewire/libpipewire-0.3-dev_1.0.5-1ubuntu3_amd64.deb  404  Not Found [IP: 52.252.75.106 80]
```

SolarLiner commented on src/lib.rs:
```diff
@@ -110,9 +110,10 @@ pub struct StreamConfig {
-    /// honoring this setting, and in future versions may provide additional buffering to ensure
-    /// it, but for now you should not make assumptions on buffer sizes based on this setting.
+    /// Range of preferential buffer sizes, in units of audio samples per channel.
+    /// The library will make a bast-effort attempt at honoring this setting, and in future versions
```
```suggestion
    /// The library will make a best-effort attempt at honoring this setting, and in future versions
```

Looks like I've made a typo writing this, if you could include the fix here that would be perfect!

SolarLiner reviewed (approved):
Looks like the CI issue was indeed transient and was in the Azure repository mirrors.

mbernat committed:
Update src/lib.rs

Co-authored-by: Nathan Graule <solarliner@gmail.com>

SolarLiner commented:
@mbernat I'll let you take care of the conflicts when you can, and I'll merge right after.

mbernat commented:
@SolarLiner Thanks, resolved!
